keystone_keycloak_backend/identity.py

```diff
@@ -20,29 +23,291 @@ def __init__(self, conf=None):
+                LOG.error(
+                    f"KEYCLOAK_ERROR: Response body: {getattr(e, 'response_body', 'No response body')}"
+                )
+                endpoint_url = f"{self.conf.keycloak.server_url}/admin/realms/{self.conf.keycloak.realm_name}/users"
+                LOG.error(f"KEYCLOAK_ERROR: Likely endpoint: {endpoint_url}")
+
+            # Handle 403 Forbidden - could be stale permissions in cached token
+            if e.response_code == 403:
+                auth_method = self._auth_method
+                auth_id = self._auth_identifier
+
+                # Try refreshing token once in case permissions were recently changed
+                if not hasattr(self, "_token_refresh_attempted"):
+                    if getattr(self.conf.keycloak, "debug", False):
+                        LOG.warning(
+                            f"KEYCLOAK_DEBUG: Got 403 for {operation_name}, "
+                            f"refreshing token in case permissions changed..."
+                        )
+                    self._token_refresh_attempted = True
+                    self._refresh_token_and_client()
+                    # Retry the operation with fresh token - get the method from the fresh instance
+                    try:
+                        operation_name = getattr(operation, "__name__", str(operation))
+                        fresh_operation = getattr(self.keycloak, operation_name)
+                        result = fresh_operation(*args, **kwargs)
+                        # Reset the flag on successful retry
+                        delattr(self, "_token_refresh_attempted")
+                        return result
+                    except keycloak_exceptions.KeycloakGetError as retry_e:
+                        if retry_e.response_code == 403:
+                            # Still 403 after refresh, it's a real permission issue
+                            delattr(self, "_token_refresh_attempted")
+                            raise Exception(
+                                self._generate_403_error_message(
+                                    operation_name, auth_method, auth_id, retry_e
+                                )
+                            )
+                        else:
+                            # Different error after retry, re-raise it
+                            delattr(self, "_token_refresh_attempted")
+                            raise
+                else:
+                    # Already attempted refresh, it's a real permission issue
+                    delattr(self, "_token_refresh_attempted")
+                    raise Exception(
+                        self._generate_403_error_message(
+                            operation_name, auth_method, auth_id, e
+                        )
+                    )
+            # Re-raise other KeycloakGetError exceptions as-is
+            raise
 
     def is_domain_aware(self):
         # TODO(mnaser): check this
         return False
 
```

Comment:
The `_keycloak_with_retry` wrapper is applied inconsistently. Some methods use it (`list_users`, `get_user`, `get_groups`, etc.) but others don't:

- `get_user_by_name` calls `self.keycloak.get_users()` directly
- `list_users_in_group` calls `self.keycloak.get_group_members()` directly  
- `authenticate` calls `self.keycloak.get_user()` directly

If the retry logic is needed for token expiry handling, it should be applied consistently to all Keycloak API calls.